apps/sms/services/js/moz_mobile_message/messaging_service.js

```diff
@@ -0,0 +1,193 @@
+/*global BridgeServiceMixin, Settings */
+'use strict';
+
+if (!('BridgeServiceMixin' in self)) {
+  importScripts('/services/js/bridge_service_mixin.js');
+}
+
+(function(exports) {
+
+const SERVICE_NAME = 'messaging-service';
+const MOBILE_MESSAGE_CLIENT_NAME = 'mozMobileMessageShim';
+
+const METHODS = Object.freeze([
+  'sendSMS', 'sendMMS', 'resendMessage', 'retrieveMMS'
+]);
+
+function ensureBridge() {
+  if (!('bridge' in self)) {
+    importScripts('/lib/bridge.js');
+  }
+}
+
+// 0 is a valid value so we need to take care to not consider it as a falsy
+// value. We want to return null for anything that's not a number or a
+// string containing a number.
+function sanitizeServiceId(serviceId) {
+  if (serviceId == null || // null or undefined
+      isNaN(+serviceId)) {
+    serviceId = null;
+  } else {
+    serviceId = +serviceId;
+  }
+
+  return serviceId;
+}
+
+function getSendOptionsFromServiceId(serviceId) {
+  var sendOpts;
+
+  if (serviceId != null && // not null, not undefined
+      Settings.hasSeveralSim()) {
+    sendOpts = { serviceId: serviceId };
+  }
+
+  return sendOpts;
+}
+
+var MessagingService = {
+  /**
+   * Initializes our service using the Service mixin's initService and
+   * requests mozMobileMessage shim as client.
+   */
+  init() {
+    ensureBridge();
+
+    this.initService();
+
+    this.mozMobileMessageClient = bridge.client(MOBILE_MESSAGE_CLIENT_NAME);
+  },
+
+  /**
+   * Send sms by using send API shim.
+   * @param {*} opts Options for sending sms contains following properties:
+   * - recipients (string or array of string): contains the list of
+   *   recipients for this message.
+   * - content (string): the message's body.
+   * - serviceId (optional long or string): the SIM serviceId we use to send
+   *   the message.
+   * @returns {Promise} Promise with array of the request result.
+   */
+  sendSMS(opts) {
+    var recipients = opts.recipients || [],
+        content = opts.content,
+        serviceId = sanitizeServiceId(opts.serviceId);
+
+    if (!Array.isArray(recipients)) {
+      recipients = [recipients];
+    }
+
+    var sendOpts = getSendOptionsFromServiceId(serviceId);
+    var results = recipients.map(
+      (recipient) => this.mozMobileMessageClient(
+        'send', recipient, content, sendOpts
+      ).then((evt) => evt, (evt) => evt)
```

Comment:
Since we don't have `finally` yet, not sure if there is other better way for this because I want aggregate all the success and error result instead of separate to resolve and reject .